File: learning_js/scripts/equality-and-truthiness.js
/* eslint-disable @typescript-eslint/no-unused-vars */
/* eslint-disable no-constant-condition */
import { log } from "console";

/* JS does NOT have:
1. chained assigments w/ declaration (i.e., let a = b = 0) where a and b are being declared OR
2. multiple assignment (a, b = 0, 1) */
{
  // BUT it DOES have expression assigments
  let a;
  if ((a = 0) < 1) {
    // assigns 0 to a and then expression evaluates to value of a
    log("assigned and expression");
  }
  // can simulate chained assignment w/o declaration this way
  let c;
  let b = (c = 10); // works because c is already declared
  // b = c = 10 after both have been declared also works
  log(a, b);
}
{
  // it ALSO DOES have destructuring assignment if you need to set multple different values
  let [a, b] = [1, 2];
  log(a, b); // 1 2
}

{
  // equality comes in two forms: normal (==) and strict (===);
  // Almost ALWAYS use === to avoid bugs
  log(1 == "1"); // true
  log(1 === "1"); // false
  // == attempts type conversion; === does not and performs reference equality checks for objects
  let a = {};
  let b = a;
  log("reference equality", a === b); // true
  // log(a === {}) would be false

  // undefined and null have reference equality with any expression/variable that is undefined or null
  a = null;
  log(a === null); // true
  // CAUTION: undefined and null are not === equal, so usually use truthiness or nullish coalescing
  // operator if checking for either (common use case)
  log(undefined === null); // false

  // there is no operator overloading in JS, so there is no way to define custom equality checks
}

{
  // all JS values evaluate to true or false in a boolean context; theses are the falsy values:
  log("falsies", 0 || -0 || NaN || "" || null || undefined || false);
  // since 0 is falsy, be careful when checking against variables that can be numbers

  // any object is truthy
  log(!!{}); // !! is quick syntax to convert to boolean (not not) instead of Boolean()
  log(!![]);
  // includes wrappers of primitives (note: 'new' constructs an object; without it, the function
  // computes the primitive conversion of the object)
  log("object wrappers", !!(new String(5) && new Number(0) && new Boolean(false)));
}

File: learning_backend/python_practice/equality-and-truthiness.py
import math


a, b = 0, 1  # multiple assignment, like JS
# falsy values in python are the same as JS (except for NaN), but also include empty basic collections
print(not ([] or () or {} or set() or range(0)))
print(
    not None
)  # None is equivalent to null/undefined in JS and the default return value of functions
print(bool(math.nan))  # NaN is truthy in python

# Value equality is ==, which calls __eq__ method, like Java.
# Reference equality is 'is', which checks if two references point to the same object
# (curiously, constants can pass an 'is' check due to memory reuse)
a = []
b = []
print(a == b)  # True
print(a is not b)  # True
# nan is not equal to itself (like JS ===)
print(math.nan != math.nan)  # True

# python supports assignment expressions (walrus operator), which performs the assignment and
# returns the value of the variable
n = 4
while (n := n - 1) >= 0:
    print(n)
